ecs-agent/ec2/ec2_metadata_client.go

```diff
@@ -94,21 +97,35 @@ type ec2MetadataClientImpl struct {
 }
 
 // NewEC2MetadataClient creates an ec2metadata client to retrieve metadata
-func NewEC2MetadataClient(client HttpClient) EC2MetadataClient {
+func NewEC2MetadataClient(client HttpClient) (EC2MetadataClient, error) {
 	if client == nil {
-		config := aws.NewConfig().WithMaxRetries(metadataRetries)
-		config.Credentials = instancecreds.GetCredentials(false)
-		return &ec2MetadataClientImpl{
-			client: ec2metadata.New(session.New(), config),
+		credentialsProvider := providers.NewInstanceCredentialsCache(
+			false,
+			providers.NewRotatingSharedCredentialsProviderV2(),
+			nil,
+		)
+		cfg, err := config.LoadDefaultConfig(
+			context.TODO(),
+			config.WithCredentialsProvider(credentialsProvider),
+			config.WithRetryer(func() aws.Retryer {
+				return retry.AddWithMaxAttempts(retry.NewStandard(), metadataRetries)
+			}),
+		)
+		if err != nil {
+			return nil, err
 		}
+
+		return &ec2MetadataClientImpl{
+			client: imds.NewFromConfig(cfg),
+		}, nil
 	} else {
-		return &ec2MetadataClientImpl{client: client}
+		return &ec2MetadataClientImpl{client: client}, nil
 	}
 }
 
 // DefaultCredentials returns the credentials associated with the instance iam role
 func (c *ec2MetadataClientImpl) DefaultCredentials() (*RoleCredentials, error) {
-	securityCredential, err := c.client.GetMetadata(SecurityCredentialsResource)
+	securityCredential, err := c.GetMetadata(SecurityCredentialsResource)
```

Comment:
[`ec2MetadataClientImpl.GetMetadata`](https://github.com/tinnywang/amazon-ecs-agent/blob/14ff880f379bc0122558bd01a6176e6f25c7be6d/ecs-agent/ec2/ec2_metadata_client.go#L178-L192) calls the underlying SDK client's [`GetMetadata`](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/feature/ec2/imds#Client.GetMetadata). It's more convenient to use when you want the metadata content as a string. Otherwise, you have to duplicate the body of [`ec2MetadataClientImpl.GetMetadata`](https://github.com/tinnywang/amazon-ecs-agent/blob/14ff880f379bc0122558bd01a6176e6f25c7be6d/ecs-agent/ec2/ec2_metadata_client.go#L178-L192) everywhere, because `c.client.GetMetadata` returns [`*GetMetadataOutput`](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/feature/ec2/imds#GetMetadataOutput).
```
	output, err := c.client.GetMetadata(context.TODO(), &imds.GetMetadataInput{
		Path: path,
	})
	if err != nil {
		return "", err
	}

	content, err := io.ReadAll(output.Content)
	if err != nil {
		return "", err
	}

	return string(content), nil
```